src/System.Linq/src/System/Linq/AnyAll.cs

```diff
@@ -15,6 +16,29 @@ public static bool Any<TSource>(this IEnumerable<TSource> source)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
             }
 
+            if (source is ICollection<TSource> collectionoft)
+            {
+                return collectionoft.Count != 0;
+            }
+            else if (source is IIListProvider<TSource> listProv)
```

Comment:
IReadOnlyCollection also defines get_Count(), but isn't caught by this waterfall. Should it be?

Reply:
Can we take this waterfall pattern out and put it in a helper method so both Count and Any can benefit from "FastCount"?

```C#
[AggressiveInlining]
private static bool TryGetFastCount(out int count)
{
    ...
}
```

That way any cleverness that is applied to one can be applied to the other.  I understand that the nullable value may itself make things less good, but perhaps there are clever things with aggressive inlining and using refs vs outs or whatever that make it work more macro-y.

Reply:
> Should it be?

We don't check for `IReadOnlyCollection<T>` in LINQ today for two reasons:
1. It's actually quite expensive, as `IReadOnlyCollection<T>` is covariant, and runtime casts for such interfaces are much more expensive.  There's a coreclr issue open tracking that getting fixed.
2. The vast majority of collection types that would benefit from `IReadOnlyCollection<T>` checks also benefit from the others, and thus we defer to (1).

There are several open issues in corefx related to this.  I'll leave this be, and we can always revisit it later when we revisit it in general across corefx.

> That way any cleverness that is applied to one can be applied to the other.

There's already one difference between Any and Count, which I commented on in the code.  That'll make such a helper more complicated, and at that point, I don't think it's worth the complexity.  This set of interface checks is very common across LINQ now, but with different behavior for each branch.